### ISSUE TITLE
Suggestions: Investigate `.{` operator for mutations

### ISSUE THREAD

**Issue by rbuckton** [rickbutton/proposal-deep-path-properties-for-record#15]
There was a strawperson many years ago for adding a `.{` ("monocle mustache") operator for assigning multiple properties to an object that might be worth investigating. IIRC, the basic idea was this:

```js
const obj = { a: 1, b: 2 };
obj.{ a: 3, c: 4 };
obj.a; // 3
obj.b; // 2
obj.c; // 4
```

This might be a useful operator to consider for replacing deep properties in records/tuples as opposed to, or in addition to, spread (`...`):

```js
const rec1 = #{ a: 1, b: #[2] };
const rec2 = rec1.{ b[0]: 3 };
rec2.a; // 1
rec2.b; // [3]
```

The question would be whether `.{` on a regular object would be mutating even if `.{` on a record/tuple would not.

**Comment by stiff** [comment]
Rather than introducing weird overload for `.` operator, which accesses property, wouldn't it be better to overload `+` for objects/maps/sets/arrays?

This would lead to intuitive code:
```javascript
obj + { a: 3, c: 4 }
obj = obj + { a: 3, c: 4 }
obj += { a: 3, c: 4 }
```

**Comment by ljharb** [comment]
@stiff no, that’s not possible, since it already has existing semantics. A new operator would be required.

**Comment by stiff** [comment]
New is ok, reusing existing with pretty much opposite function to previous is not.

**Comment by ljharb** [comment]
By "new" i mean "was a syntax error previously". `.{` is new, `+` with objects is not.

**Comment by stiff** [comment]
Btw, the `.{}` and `.[]` are nice names for [pick](https://ramdajs.com/docs/#pick) and values_at ([props](https://ramdajs.com/docs/#props)). Then update should be something like `!{}`.

**Comment by hax** [comment]
>since it already has existing semantics.

@ljharb I think overloading `+` for tuple/record is possible, just like how we overload `+` for bigints. So `#[1, 2] + #[3, 4] === #[1,2,3,4]`, `#{a:1} + #{b:2} === #{a:1, b:2}` could work. I'm not sure whether such overloading is good or bad, but I guess that we'd better throw TypeError for any `+` operation when operand is tuple/record for now, so we keep the overloading possibility in the future.
